modules/partition-replicator/src/main/java/org/apache/ignite/internal/partition/replicator/PartitionReplicaLifecycleManager.java

```diff
@@ -0,0 +1,425 @@
+
+                        return realAssignments;
+                    });
+        });
+    }
+
+    /**
+     * Check if the zone already has assignments in the meta storage locally.
+     * So, it means, that it is a recovery process and we should use the meta storage local assignments instead of calculation
+     * of the new ones.
+     */
+    private CompletableFuture<List<Assignments>> getOrCreateAssignments(
+            CatalogZoneDescriptor zoneDescriptor,
+            long causalityToken,
+            int catalogVersion
+    ) {
+        CompletableFuture<List<Assignments>> assignmentsFuture;
+
+        if (zonePartitionAssignmentsGetLocally(metaStorageMgr, zoneDescriptor.id(), 0, causalityToken) != null) {
+            assignmentsFuture = completedFuture(
+                    zoneAssignmentsGetLocally(metaStorageMgr, zoneDescriptor.id(), zoneDescriptor.partitions(), causalityToken));
+        } else {
+            assignmentsFuture = distributionZoneMgr.dataNodes(causalityToken, catalogVersion, zoneDescriptor.id())
+                    .thenApply(dataNodes -> AffinityUtils.calculateAssignments(
+                            dataNodes,
+                            zoneDescriptor.partitions(),
+                            zoneDescriptor.replicas()
+                    ).stream().map(Assignments::of).collect(toList()));
+
+            assignmentsFuture.thenAccept(assignmentsList -> LOG.info(
+                    "Assignments calculated from data nodes [zone={}, zoneId={}, assignments={}, revision={}]",
+                    zoneDescriptor.name(),
+                    zoneDescriptor.id(),
+                    Assignments.assignmentListToString(assignmentsList),
+                    causalityToken
+            ));
+        }
+
+        return assignmentsFuture;
+    }
+
+    @Override
+    public CompletableFuture<Void> stopAsync(ComponentContext componentContext) {
+        if (!ENABLED) {
+            return nullCompletedFuture();
+        }
+
+        return nullCompletedFuture();
+    }
+
+    /**
+     * Stops all resources associated with a given partition, like replicas and partition trackers.
+     *
+     * @param zonePartitionId Partition ID.
+     * @return Future that will be completed after all resources have been closed.
+     */
+    private CompletableFuture<?> stopPartition(ReplicationGroupId zonePartitionId) {
+        CompletableFuture<Boolean> stopReplicaFuture;
+
+        try {
+            stopReplicaFuture = replicaMgr.stopReplica(zonePartitionId);
+        } catch (NodeStoppingException e) {
+            // No-op.
+            stopReplicaFuture = falseCompletedFuture();
+        }
+
+        return stopReplicaFuture;
+    }
+
+    /**
+     * Stops resources that are related to provided zones.
+     *
+     * @param zones Zones to stop.
+     */
+    private void cleanUpZoneResources(Set<ReplicationGroupId> zones) {
+        var futures = new ArrayList<CompletableFuture<Void>>(zones.size());
```

Comment:
By the way, because of the poor naming, the code is messy. 

- futures list size is partitionsSize and not zones. Only one element is added to the list, which is allOf(partitions) line 409.
- `LOG.error("Unable to stop zone");` is basically an equivalent of `LOG.error("Unable to clean zone resources", e);` in particular case.

In Table manager there are two loops: outer for tables and inner for table partitions. In PRLM partitions are first class citizens.

Reply:
@alievmirza FYI^